laurent22/joplin
# Desktop: Fixes #8535: Fix CodeMirror context menu not containing correct items

# Summary
Should partially fix #8535.


## The issue
The default context menu shown in [`electron-context-menu.js`](https://github.com/laurent22/joplin/blob/087b20698218675c1e92b88be0e95df734d0e2c2/packages/app-desktop/services/electron-context-menu.js#L236) was conflicting with the context menu shown in `CodeMirror.tsx`.

With the upgrade to Electron 25, the reported `inputFieldType` for a `textarea` changed from `none` to `plainText` when spellcheck is disabled. This broke logic in `bridge.ts` that previously prevented the default context menu from being shown.

Additionally, this broke the logic that caused a custom context menu to be shown for the CodeMirror editor.

This pull request replaces the logic in `bridge.ts` that originally prevented the default context menu from being displayed with a check for whether the context menu event has had `.preventDefault` called by another event listener.

The listener in `CodeMirror.tsx` is now registered with `prependListener` so that it is called before the default context menu listener. The `CodeMirror.tsx` listener additionally calls `.preventDefault` to prevent the default context menu from being displayed.

# Testing

To verify that the updated CodeMirror context menu is working correctly,
1. Install a plugin that adds items to the context menu (e.g. Rich Markdown)
2. Restart Joplin
3. Using the CodeMirror editor, add a checklist
4. Right click on an item in the checklist

In step 4, a "Toggle Checkbox" item should be visible.

5. Enable spellcheck in the markdown editor
6. Add a misspelled checklist item
7. Right-click on the misspelling and correct it
8. Right-click on a note
9. Right-click in the rich text editor

This has been tested on Ubuntu 23.10.
